src/Features/ExternalAccess/Copilot/Internal/Completion/CSharpContextProviderService.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.ExternalAccess.Copilot.Completion;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Copilot.Internal.Completion;
+
+[Shared]
+[Export(typeof(ICSharpCopilotContextProviderService))]
+internal sealed class CSharpContextProviderService : ICSharpCopilotContextProviderService
+{
+    // Exposed for testing
+    public ImmutableArray<IContextProvider> Providers { get; }
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CSharpContextProviderService([ImportMany] IEnumerable<IContextProvider> providers)
+    {
+        Providers = providers.ToImmutableArray();
+    }
+
+    public async IAsyncEnumerable<IContextItem> GetContextItemsAsync(Document document, int position, IReadOnlyDictionary<string, object> activeExperiments, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        var queue = new AsyncQueue<IContextItem>();
+        var tasks = this.Providers.Select(provider => Task.Run(async () =>
+        {
+            try
+            {
+                await provider.ProvideContextItemsAsync(document, position, activeExperiments, ProvideItemsAsync, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception exception) when (FatalError.ReportAndCatchUnlessCanceled(exception, ErrorSeverity.General))
+            {
+            }
+        },
+        cancellationToken));
+
+        // Let all providers run in parallel in the background, so we can steam results as they come in.
+        // Complete the queue when all providers are done.
+        _ = Task.WhenAll(tasks)
```

Comment:
we have ProducerConsumer which can parallelize, without unbounded task spawing.

Reply:
is this code running in proc in vs?  if so, we do not do parallel fork out as it can significantly degrade teh VS experience (lots of garbage and cpu contending with VS threads).  If this is in OOP, this is ok, but we prefer ProducerConsumer as it utilizes cores more effectively and ensures consistent progress.   

For example, if you have 8 cores, it will be running 8 tasks, reporting results as they proceed, and starting new tasks as they complete.

If you use do Task.WhenAll, then you could be starting up N tasks (where n is very large), all of which make tiny progress at a time, reporting nothing, until an avalanche of completions at the end.